cmd/pinniped-proxy/src/pinniped.rs

```diff
@@ -239,88 +166,82 @@ async fn call_pinniped_exchange(
+        TOKEN_REQUEST_KIND,
+    );
+    let ar = ApiResource::from_gvk(&gvk);
 
-/// call_pinniped_exchange returns the resulting TokenCredentialRequest with Status after requesting a token credential exchange.
-// TODO(agamez): remove this duplicated code and use a generic dynamic approach when possible.
-// see PR comments: https://github.com/kubeapps/kubeapps/pull/2515
-async fn call_pinniped_exchange_tmc(
-    authorization: &str,
-    k8s_api_server_url: &str,
-    k8s_api_ca_cert_data: &[u8],
-) -> Result<TokenCredentialRequestTMC> {
-    let pinniped_namespace = env::var(DEFAULT_PINNIPED_NAMESPACE)?;
-    let client = get_client_config(
-        k8s_api_server_url,
-        k8s_api_ca_cert_data,
-        pinniped_namespace.clone(),
-    )?;
-    let auth_token = match authorization.to_string().strip_prefix("Bearer ") {
-        Some(a) => a.to_string(),
-        None => authorization.to_string(),
-    };
-    let token_creds: Api<TokenCredentialRequestTMC> = Api::all(client.clone());
-    let cred_request = TokenCredentialRequestTMC::new(
-        "",
-        TokenCredentialRequestTMCSpec {
+    // create request
+    let cred_data = TokenCredentialRequest {
+        spec: TokenCredentialRequestSpec {
             token: Some(auth_token),
             authenticator: corev1::TypedLocalObjectReference {
-                name: env::var(DEFAULT_PINNIPED_AUTHENTICATOR_NAME).with_context(|| {
-                    format!("error retrieving {}", DEFAULT_PINNIPED_AUTHENTICATOR_NAME)
-                })?,
-                kind: env::var(DEFAULT_PINNIPED_AUTHENTICATOR_TYPE).with_context(|| {
-                    format!("error retrieving {}", DEFAULT_PINNIPED_AUTHENTICATOR_TYPE)
-                })?,
+                name: pinniped_auth_name.clone(),
+                kind: pinniped_auth_type.clone(),
                 api_group: Some(get_pinniped_authenticator_api_group().into()),
             },
         },
-    );
-
+        status: None,
+    };
+    let cred_request = DynamicObject::new("", &ar)
+        .within(&pinniped_namespace)
+        .data(serde_json::to_value(cred_data)?);
     debug!("{}", serde_json::to_string(&cred_request).unwrap());
-    match token_creds
+
+    // token credential request invocation
+    // we start first as a cluster-based call. if this call fails with a NotFound error, it is
+    // an indication we are on an old version which was namespace-based. we thus fallback to
+    // a namespace-based call.
+    let token_creds_all: Api<DynamicObject> = Api::all_with(client.clone(), &ar);
+    match token_creds_all
         .create(&PostParams::default(), &cred_request)
         .await
     {
-        Ok(o) => Ok(o),
+        Ok(o) => Ok(serde_json::from_value(o.data)?),
+        Err(kube::Error::Api(e)) => {
+            if e.reason == "NotFound" {
+                let token_creds_ns: Api<DynamicObject> =
+                    Api::namespaced_with(client.clone(), &pinniped_namespace, &ar);
```

Comment:
like adding a function "isFallbackPossible(e: Error)" so we can test that one function?
we can add a test that ensures we are not regressing our own logic, but this will not ensure that the library logic does not change.
if you think adding such method is useful, i can refactor the code to add it.

Reply:
Yep, perhaps refactoring it is not really worthy,  but adding a test for that functionality would be great, IMO

Reply:
ok, i'll look into it